datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -47,6 +47,80 @@ where
         }
     }
 
+    fn traces(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+        addresses: Vec<H160>,
+    ) -> impl Future<Item = Vec<Trace>, Error = Error> {
+        let eth = self.clone();
+        let logger = logger.to_owned();
+
+        retry("trace_filter RPC call", &logger)
+            .no_limit()
+            .timeout_secs(60)
+            .run(move || {
+                let trace_filter: TraceFilter = match addresses.len() {
+                    0 => TraceFilterBuilder::default()
+                        .from_block(from.into())
+                        .to_block(to.into())
+                        .build(),
+                    _ => TraceFilterBuilder::default()
+                        .from_block(from.into())
+                        .to_block(to.into())
+                        .to_address(addresses.clone())
+                        .build(),
+                };
+
+                let logger = logger.clone();
+                let logger_1 = logger.clone();
```

Comment:
Throughout the code base we've been suffixing clones with their purpose (e.g. `logger_for_err`) instead of enumerating them.